src/python/pants/engine/intrinsics.py

```diff
@@ -107,6 +110,20 @@ async def execute_process(
     return await native_engine.execute_process(process, process_execution_environment)
 
 
+@rule
+async def execute_process_with_retry(req: ProcessWithRetries) -> ProcessResultWithRetries:
```

Comment:
Technically not an intrinsic but it's a helper for working with an intrinsic, so it seems OK to have it here. There are other similar rules in this file. Or we can make an ad-hoc file for it I suppose.